### PR TITLE
Generalize inference of repository configuration

### PR DESCRIPTION
**Pull request by mjtko**
Previously ActiveRecord was required to automatically infer the database
configuration.  This has been generalized and now also supports
DataMapper, as well as attempting to fall back to the environment
variable DATABASE_URL if available.

### CODE REVIEW

File: lib/rom/rails/railtie.rb [hunk 1]
```diff
@@ -79,9 +79,12 @@ def setup
...
 
       def infer_default_repository
-        return unless defined?(ActiveRecord)
-        spec = ActiveRecord::Configuration.call
-        [:sql, spec[:uri], spec[:options]]
+        if RepositoryConfiguration.inferrable?
+          spec = RepositoryConfiguration.infer
+          [:sql, spec[:uri], spec[:options]]
+        elsif ENV['DATABASE_URL']
+          [:sql, ENV['DATABASE_URL']]
```
Comment (aflatter) [inline lib/rom/rails/railtie.rb]: This should be part of the helper class too, it's yet another source of configuration in addition to DM and AR.

Comment (mjtko) [inline lib/rom/rails/railtie.rb]: Good point, I'll move it.

**Comment by aflatter** [comment]
Hey @mjtko,

thanks for that PR. I'm a bit sceptic about DM and AR using the exact same type of options hash. DM supports much more than just SQL, so from my point of view it deserves its own "inferrer class". If you're not sure about that either, maybe @solnic can chime in.

**Comment by mjtko** [comment]
@aflatter Yup, fair enough.  Perhaps some kind of "does this look like the right kind of options Hash, raise if not" approach would cover (_plucks figure out of thin air_) 90% of DM use?  I would imagine when used with Rails it's pretty regularly configured in the conventional way.

Feels like overkill to cover everything that DM might do -- this mainly arose from needing to get something up and running with ROM+DM+Rails (w/o AR).

File: lib/rom/rails/railtie.rb (hunk 1 above)
Comment (mjtko) [inline lib/rom/rails/railtie.rb]: I'm going to hold off on the relocation for the moment.  I think we need to decide how many helper classes we need beforehand.

If we're having a separate helper for each source we can infer the configuration from, then this probably deserves its own (simple) one.  Currently it's not a good fit for the `RepositoryConfiguration` class which is specifically dealing with options hashes.

**Comment by solnic** [comment]
I believe we should focus on supporting common cases, otherwise it'd be a partially wasted effort.

**Comment by mjtko** [comment]
This seems related to rom-rb/rom#108 too.  Perhaps we should attempt to configure ROM with the following approaches (first discovered to give a workable configuration gets priority):

1. An initializer
2. Read and interpret `config/rom.yml` (rom-rb/rom#108)
3. Read and interpret `config/database.yml`
4. Infer from ActiveRecord configuration (if available) -- this is likely to be superceded by the above, but I guess there are some cases where AR is directly configured without a `config/database.yml`
5. infer from RDBMS-style DataMapper configuration (if available)
6. infer from DATABASE_URL (if available)

If none of those approaches work, raise an error.

How does that sound?

**Comment by mjtko** [comment]
I'm going to close this in favour of a new PR.

**Pull request closed without merging**